Improve Preference Filtering

colin-grant-work opened the pull request:
Signed-off-by: Colin Grant <colin.grant@ericsson.com>

<!--
Thank you for your Pull Request. Please provide a description and review
the requirements below.

Contributors guide: https://github.com/theia-ide/theia/blob/master/CONTRIBUTING.md
-->

#### What it does
<!-- Include relevant issues and describe how they are addressed. -->

This PR aims to make the filtering of nodes in the preference widget more performant, and to smooth the way for feature enhancements like #8252.

At the moment, filtering and scope changing actually modify the `root` of the TreeModel, rather than maintaining the same root and conditionally rendering tree nodes based on various criteria. The performance fix in #7936 improves the situation, but at the cost of some flexibility in scoping preferences. This PR moves that functionality into a modified `TreeModel` implementation and uses an iterative strategy parallel to the `doUpdateRows` function in the `TreeWidget` to determine which nodes to display. This will make filtering more performant as well as more flexible in response to changing demands.

In addition, this PR refactors away some of the not-well-encapsulated code from #7105 by eliminating the `PreferenceEventService` and locating emitters and state tracking in the classes logically responsible for them.

Also fixes #8689

#### How to test
<!-- Explain how a reviewer can reproduce a bug, test new functionality or verify performance improvements. -->

It should retain the functionality of #7105, #7773, #7764, and #7936.

In addition:
- Switching from User/Workspace to Folder scope or vice versa should be faster.
1. Open a multi-root or saved workspace.
1. Open preferences widget.
1. Switch between User scope and Folder scope.
1. Observe that the transition is quick, with little visible delay.

- Filter performance is improved.
1. Open preference widget
1. Type a text query in the search bar.
1. Observe that the filter is smooth, with little delay.

- Node visibility is preserved better.
1. Open preferences widget.
1. Scroll in editor until some preference in the section `css.lint` is at the top of the visible window.
1. Switch to folder scope.
1. Observe that the same preference is still visible. (Formerly this would have caused a scroll to top.)
1. Type the name of the preference in the filter bar.
1. Observe that the preference is still at the top of the screen. (Formerly a filter would have caused a scroll to top.)

- User and Workspace scopes are distinguished properly
1. Follow the steps outlined in #8689 
2. Observe that the transition between User and Workspace scopes is managed correctly.

Further:
- Filter behavior can now be customized by modifying or overriding the `passesCurrentFilters()` method of the `PreferenceTreeModel`. This will mean that implementing enhancements like #8252 is straightforward and performant.

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

colin-grant-work commented:
@kenneth-marut-work, would you be willing to take a look at this?

kenneth-marut-work commented:
In the third set of steps, I can confirm that typing the name of a preference (`css.lint` for example) scrolls to the correct preference. However, deleting the search query does not revert the editor back to its default (scroll-to-top) position, instead I am seeing a quick flicker and the editor returns back to the previously searched query. 
In VSCode, if a query is searched for, and a preference is selected, clearing the search query reverts the editor to its default state.

kenneth-marut-work commented on packages/preferences/src/browser/preference-tree-model.ts:
```diff
@@ -0,0 +1,184 @@
+
+    protected lastSearchedFuzzy: string = '';
+    protected lastSearchedLiteral: string = '';
+    protected _currentScope: number = parseInt(Preference.DEFAULT_SCOPE.scope);
```
For parsing scope, you've used `Number` in a couple places vs. `parseInt` here, just FYI

colin-grant-work replied:
There's certainly no doubt you were paying attention 😉 I've gone with `Number` here, for consistency.

kenneth-marut-work commented on packages/preferences/src/browser/views/preference-widget-bindings.ts:
```diff
@@ -24,8 +24,11 @@ import { PreferencesSearchbarWidget } from './preference-searchbar-widget';
+import { PreferenceTreeModel } from '../preference-tree-model';
 
 export function bindPreferencesWidgets(bind: interfaces.Bind): void {
+    let preferencesTreeContainer: Container;
```
This issue can potentially be avoided by creating a subcontainer for PreferenceWidget and binding all preference sub-widgets/elements within that container

colin-grant-work replied:
Good idea 😃

colin-grant-work replied:
I've implemented the bindings in this way, and it seems everything is working and the bindings are cleaner.

colin-grant-work commented:
Thanks for the suggestions, @kenneth-marut-work. I have adjusted the editor to scroll to top when the text filter is cleared, as you indicated, and cleaned up the code you commented on.

kenneth-marut-work commented on packages/preferences/src/browser/views/preference-editor-widget.tsx:
```diff
@@ -177,23 +205,18 @@ export class PreferencesEditorWidget extends ReactWidget {
+                const el = document.getElementById(`${nodeID}-editor`);
+                if (el) {
+                    // Timeout to allow render cycle to finish.
+                    // setTimeout(() => el.scrollIntoView());
```
Another quick catch, are these comments intentional?

colin-grant-work replied:
One yes, one no :). Thanks!

colin-grant-work commented on packages/preferences/src/browser/views/components/single-preference-wrapper.tsx:
```diff
@@ -139,22 +139,21 @@ export class SinglePreferenceWrapper extends React.Component<SinglePreferenceWra
-
-                    if (matchingValue !== undefined) {
+                    const otherValue = this.getValueByScope(info, otherScope);
+                    if (otherValue !== undefined && otherValue !== defaultValue) {
```
@vince-fugnitto, I changed this line to check that it `!== defaultValue` because I was getting a bug where if I changed the value in `User` and not in `Workspace`, I was still getting a message in `User` saying that it was modified in `Workspace`.

colin-grant-work replied:
I think the issue there was that it's possible to explicitly set the value of the preference to be the same as the default, in which case the `!== undefined` check would pass, but it still wouldn't really have been modified away from the default.

akosyakov commented on packages/preferences/src/browser/preference-tree-model.ts:
```diff
@@ -0,0 +1,185 @@
+    protected lastSearchedLiteral: string = '';
+    protected _currentScope: number = Number(Preference.DEFAULT_SCOPE.scope);
+    protected _isFiltered: boolean = false;
+    protected _currentRows: Map<string, PreferenceTreeNodeRow> = new Map();
```
The tree model should not had such representation. Why do you need it here? Flat rows should be computed in the widget based on nodes' attributes like `visible`.

colin-grant-work replied:
Because this tree - and the rules about what should and shouldn't be displayed - are common to the `PreferenceTreeWidget` and `PreferenceEditorWidget` it's more convenient and more efficient to simply share that data in a location accessible to both widgets. Similarly, scroll actions in the `PereferenceEditorWidget` affect the expansion and selection states of the `PreferenceTreeWidget`, so sharing the model between the two of them allows the `PreferenceEditorWidget` to implement that logic while the `PreferenceTreeWidget` is able to simply react to the model's selection and expansion events normally.

colin-grant-work commented:
@vince-fugnitto, @marechal-p, I've brought this up to date with `master` and it looks like the functionality is working well. If you wouldn't mind taking a look, I'd appreciate it - and I think it'll improve the preference system a fair bit 😄.

vince-fugnitto commented on packages/preferences/src/browser/util/preference-tree-generator.ts:
```diff
@@ -46,29 +55,36 @@ export class PreferenceTreeGenerator {
                 const leafNode = this.createLeafNode(propertyName, parent);
-                parent.leaves.push(leafNode);
+                CompositeTreeNode.addChild(parent, leafNode);
+                // parent.leaves.push(leafNode);
```
leftover?

colin-grant-work replied:
Good catch; I've removed it.

vince-fugnitto reviewed (approved):
@colin-grant-work I verified the changes with the provided use-cases and it works as described :+1: 
- preferences are generally much more responsive and seamless
- switching between preference scopes works well (user, workspace and folder)
- searching is more responsive, switching scopes with a search term present is also quick

marechal-p reviewed (commented):
LGTM, although I wasn't able to notice a difference with performance. Is there something to specifically do or is my computer just fast enough to make the diff negligible?

colin-grant-work commented:
Yeah, on fast machines it isn't really noticeable - but the bugfix for workspace-scoped vs. user-scoped preferences still applies. The main difference is that in the old code, the entire tree was reconstructed for any update, and now it's just being filtered in a style similar to the way nodes are hidden or shown on expansion.
